persist/sqlite/consensus.go

```diff
@@ -621,6 +645,164 @@ func (ut *updateTx) RevertEvents(index types.ChainIndex) error {
+		}
+
+		balance := balances[addrID]
+		if matured {
+			balance.Siacoins = balance.Siacoins.Add(value)
+		} else {
+			balance.ImmatureSiacoins = balance.ImmatureSiacoins.Add(value)
+		}
+		balances[addrID] = balance
+	}
+	return balances, rows.Err()
+}
+
+func (ut *updateTx) getOrphanedSiafundBalance(indexID int64) (map[int64]uint64, error) {
+	const query = `SELECT address_id, siafund_value
+FROM siafund_elements
+WHERE chain_index_id=$1`
+	rows, err := ut.tx.Query(query, indexID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to query siafund elements: %w", err)
+	}
+	defer rows.Close()
+
+	balances := make(map[int64]uint64)
+	for rows.Next() {
+		var addrID int64
+		var value uint64
+
+		if err := rows.Scan(&addrID, &value); err != nil {
+			return nil, fmt.Errorf("failed to scan siafund element: %w", err)
+		}
+		balances[addrID] += value
+	}
+	return balances, rows.Err()
+}
+
+func (ut *updateTx) getOrphanedIndexes(index types.ChainIndex) (orphaned []int64, err error) {
+	rows, err := ut.tx.Query(`SELECT id FROM chain_indices WHERE height=$1 AND block_id<>$2`, index.Height, encode(index.ID))
+	if err != nil {
+		return nil, fmt.Errorf("failed to query orphans: %w", err)
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var indexID int64
+		if err := rows.Scan(&indexID); err != nil {
+			return nil, fmt.Errorf("failed to scan orphan: %w", err)
+		}
+		orphaned = append(orphaned, indexID)
+	}
+	return orphaned, rows.Err()
+}
+
+// RevertOrphans reverts any chain indices that were orphaned by the given index
+func (ut *updateTx) RevertOrphans(index types.ChainIndex) (reverted []types.BlockID, err error) {
+	log := ut.tx.log.Named("RevertOrphans").With(zap.Uint64("height", index.Height), zap.Stringer("applied", index.ID))
+
+	orphaned, err := ut.getOrphanedIndexes(index)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get orphaned indexes: %w", err)
+	}
+
+	if len(orphaned) == 0 {
+		return nil, nil
+	}
+
+	var revertedBalance map[int64]wallet.Balance
+	for _, id := range orphaned {
+		// revert siacoin balances
+		siacoins, err := ut.getOrphanedSiacoinBalance(id)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get orphaned siacoin elements: %w", err)
+		}
+
+		// revert siafund balances
+		siafunds, err := ut.getOrphanedSiafundBalance(id)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get orphaned siafund elements: %w", err)
+		}
+
+		for addr, balance := range siafunds {
+			b := siacoins[addr]
+			b.Siafunds = balance
+			siacoins[addr] = b
+		}
+		revertedBalance = siacoins
+	}
+
+	getBalanceStmt, err := ut.tx.Prepare(`SELECT siacoin_balance, immature_siacoin_balance, siafund_balance FROM sia_addresses WHERE id=$1`)
+	if err != nil {
+		return nil, fmt.Errorf("failed to prepare balance statement: %w", err)
+	}
+	defer getBalanceStmt.Close()
+
+	updateBalanceStmt, err := ut.tx.Prepare(`UPDATE sia_addresses SET siacoin_balance=$1, immature_siacoin_balance=$2, siafund_balance=$3 WHERE id=$4`)
```

Comment:
Do we expect that many outputs per address and that many addresses that fetching the sums from the database on the fly and/or caching the balances only in memory is no longer feasible?

Reply:
The void address has a couple million utxos. Other than that, it’s possible for exchange wallets to have a significant number of both. Precalculating and storing the value is cheaper.